### ISSUE TITLE
Lockfile maintenance PR closed rather than updated

### ISSUE THREAD

**Issue by edmorley** [renovateapp/renovate#1422]
Hi!

A lockfile maintenance PR was opened (mozilla/treeherder#3157), however I then merged another Renovate PR first (mozilla/treeherder#3156), which caused GitHub to show "this PR has conflicts" message on the lockfile maintenance PR. However shortly after that PR was closed by Renovate, rather than being rebased/recreated as I would have expected.

**Comment by rarkins** [comment]
@edmorley it's intended behaviour, but maybe it can be reconsidered. Renovate deliberately closes the lock file maintenance PR as soon as it gets conflicts. The idea of lock file maintenance was to refresh it in case it got old. But if another PR has updated it recently.. I assumed we don't really need to be refreshing it. Probably better though if we still recreate it and check if there still remains any changes before closing it.

**Comment by edmorley** [comment]
Ah that makes sense. Unfortunately `yarn upgrade <package>` only touches the parts of `yarn.lock` that were related to the package in question (and in fact due to bugs, doesn't even update all instances of that package anyway, see https://github.com/yarnpkg/yarn/issues/3967#issuecomment-359299894). As such, it's very unlikely that merging another PR will ever perform all of the updates that the lockfile refresh would have performed.

Perhaps a better heuristic would be:
```
if conflicts:
    recreate changes locally
    if empty commit/diff:
        close existing PR
    else:
        push new commit to PR
```

...that way if the other PR had really updated the lockfile properly, then the PR will still get closed, but otherwise it will be updated as desired.

**Comment by rarkins** [comment]
Yes, I agree with that logic.

BTW in the earliest days of `yarn.lock`, I was under the impression that any change to the file would result in a complete "refreshing" of the contents. Maybe I was wrong, maybe they fixed it.. but today's functionality is better and as you described. As such, I agree with your pseudocode above.

**Comment by rarkins** [comment]
Slightly expanded logic:

1. If branch does not exist - and we're within schedule => refresh the lockfile and create PR if it's changed
2. If branch exists and does not have conflicts => leave it, don't attempt to refresh it again
3. If branch exists and has conflicts
   * If no changes left (diff is zero): close existing PR
   * If changes, then force push new commit over existing

The problem is, behaviour is somewhat inconsistent. If there's never any conflicts then the lock file will stay as it is indefinitely, even if it is itself out of date. But if there *is* a conflict - e.g. from merging another Renovate PR - then in that case you *do* get the very latest.

There are downsides to both the bot owner as well as the repository users if you churn the lock file maintenance PR non-stop:
- For the bot owner, it means regenerating the lockfile "just in case" every single time it runs, which is usually 1+ times per hour
- For the repository owner, it means the branch/commit will get pushed to every single time there *is* an update, which could be 10+ times per day if you have a large tree. The PR will sit at or near the top of the PR list non stop and CI will be running non-stop

**Comment by rarkins** [comment]
@edmorley what do you think about the following approach?

1. Normally, lock file maintenance PRs are not updated/refreshed
2. If lock file maintenance get a conflict, they will be *recreated*  and completely refreshed
3. If users want to manually trigger a refresh, they need to close the existing PR to have Renovate recreate on next run

**Comment by edmorley** [comment]
Our repo was using the default though - and was still being updated multiple times. "Before 3am Monday" is a time window then, over which multiple runs can occur, rather than a single point in time when a cron job runs - right?

**Comment by rarkins** [comment]
Right. So in the original implementation it might have created it in the first hour and retried it twice more after. More times if someone committed to master in those hours and triggered an on-demand run! Now, it creates it only once and doesn't retry. The logic mistake was that it was saying "don't retry it unless the branch falls behind master" so after it did, it would keep retrying every run!

**Comment by edmorley** [comment]
> If users want to manually trigger a refresh, they need to close the existing PR to have Renovate recreate on next run

I just tried doing this for our latest lockfile maintenance PR (which was closed by the bots for conflicts), however it didn't result in a new PR?
https://github.com/mozilla/treeherder/pull/3193

**Comment by rarkins** [comment]
@edmorley I think that's because of the schedule (`":maintainLockFilesWeekly"`), i.e. it only tries lock files weekly, not every time.
